### PR TITLE
WIP :[Bug]: GetTransactionChain query timeouts.| Resolves  #180

### LINKED ISSUE #180: [Bug]: GetTransactionChain query timeouts after sometime
### What happened?

After a long run of cycles, more than 1 hour with self-repair cycle of 1min interval, the database requests `GetTransactionChain` timeouts, maybe caused by the size of a chain to pull.

### What is expected ?

We should be able to read a transaction chain without timeout whatever the size.

### PR DESCRIPTION
**Pull request by apoorv-2204**
# Description
https://misc.archethic.net/s/Frfl7j2oZ

Fixes #180

Issues :

 
``` 1/3 Message Transmission Timeout```
**Time for the db to fetch everything and time to be transmitted to the network Crosses the timeout limitation of p2p.**

 When the responder node  receives   GetTransaction type Request()  , it has timeout of 5 seconds to run the query on db and transmit the data back to the sender node in parts . 
 -  Time_db - Time taken to Run the query for specific address  and give back result set (with paging_state).
 -  time_parse - Time taken to parse the result set in to [ArchEthic.TransactionChain.Transaction.t()] .
 -  Time_upload - Time taken to transmit/upload (ArchEthic.TransactionChain.Transaction.t().TransactionList)  
     to the Sender Node.
```  Constraint : time_db + time_parse + time_upload  + time_parse(reason_unknow) < 5seconds```

> Proposed Solution :  Reduce the rows in the db query result set with paging state.

2/3
```DB connection time-out```
Xandra establishes connection with cassandra/scylla db with underlying library  [db_connection](https://hex.pm/packages/db_connection) library . 

1. When there is struggle for resources , the DB_Connection might drop of the connection from the connection pooling, resulting in loss of paging state , result set , cached connection in connection pooling etc . 

2. db-Connection Backoff : 

> Proposed Solution :  Persist the query and the paging state.

3/3
``` undetermined```

# Tasks
- [x]  flow of the issue (method flow , from start to end)
- [x]  Change Request & Response to accommodate paging state
- [x]  Re implement : Message.process for get txn chain
- [ ]   incorporate changes to accept new %GetTransactionChain{} & %TransactionList{}
- [x]  A recursive method That requests %Get Transaction Chain.
- [ ]  Implement new query (paging-state)
- [ ]  Handle edge cases
- [ ]  Implement changes in encode & decode .

# How Has This Been Tested?
- [ ] unit tests
- [ ] doc tests
- [ ] integration tests
- [ ] add specs

# Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: Schema_patch [hunk 1]
```diff
@@ -0,0 +1,178 @@
+commit 58458986e40e3609086d1f4911089dcdbfce91fa
```
Comment (samuel-uniris) [inline Schema_patch:1]: I think it's an oversight  ;)

File: lib/archethic/db/cassandra_impl.ex [hunk 2]
```diff
@@ -74,52 +74,97 @@ defmodule ArchEthic.DB.CassandraImpl do
     end
   end
 
+  # @impl DB
```
Comment (samuel-uniris) [inline lib/archethic/db/cassandra_impl.ex:77]: The same goes here

File: lib/archethic/db/cassandra_impl.ex [hunk 3]
```diff
@@ -74,52 +74,97 @@ defmodule ArchEthic.DB.CassandraImpl do
...
 
     :telemetry.execute([:archethic, :db], %{duration: System.monotonic_time() - start}, %{
       query: "get_transaction_chain"
     })
 
-    chain
+    {chain, more?, paging_state}
   end
 
+  defp get_transaction_chain_query(address, after_time: after_time, page: _current_page_state)
```
Comment (samuel-uniris) [inline lib/archethic/db/cassandra_impl.ex:147]: You can directly pattern match `after_time: nil` without the guard.

File: lib/archethic/db/cassandra_impl.ex [hunk 4]
```diff
@@ -298,22 +343,21 @@ defmodule ArchEthic.DB.CassandraImpl do
...
       tx_timestamp
     ])
 
     :ok
   end
 
-  defp bucket_from_date(%DateTime{month: month}) do
-    div(month + 2, 3)
-  end
+  # defp bucket_from_date(%DateTime{month: month}) do
```
Comment (samuel-uniris) [inline lib/archethic/db/cassandra_impl.ex:358]: You can remove this as well

File: lib/archethic/p2p/message.ex [hunk 5]
```diff
@@ -996,21 +996,26 @@ defmodule ArchEthic.P2P.Message do
...
-    }
+    # new_page_state contains the  page number of data being sent
+    %TransactionList{transactions: chain, page: new_page_state, more?: more?}
   end
 
-  def process(%GetTransactionChain{address: tx_address, after: nil}) do
-    %TransactionList{
-      transactions: TransactionChain.get(tx_address)
-    }
+  def process(%GetTransactionChain{address: tx_address, after: nil, page: current_page_state}) do
```
Comment (samuel-uniris) [inline lib/archethic/p2p/message.ex:1013]: Why do multiple calls ?
Is not the TransactionChain.get or the DB impl to manage the switch based on the options ?

File: lib/archethic/replication/transaction_context.ex [hunk 6]
```diff
@@ -42,19 +42,132 @@ defmodule ArchEthic.Replication.TransactionContext do
   defp do_fetch_transaction_chain([node | rest], address, _prev_result) do
     case P2P.send_message(node, %GetTransactionChain{address: address}) do
       {:ok, %TransactionList{transactions: []}} ->
+        # if the selected node from rest dont have the txn list / node is unavailable
+        #   then try to fetch frpm rest of  other nodes
```
Comment (samuel-uniris) [inline lib/archethic/replication/transaction_context.ex:46]: Some typo I guess

File: lib/archethic/replication/transaction_context.ex [hunk 7]
```diff
@@ -42,19 +42,132 @@ defmodule ArchEthic.Replication.TransactionContext do
...
 
   defp do_fetch_transaction_chain([], _address, nil), do: raise("Cannot fetch transaction chain")
   defp do_fetch_transaction_chain([], _address, prev_result), do: prev_result
 
+  # ===========================================================================================
+
+  @doc """
+  Finds the closest node and starts the process of replication of complete TXNCHAIN if exists
+  """
+  def fetch_complete_transaction_chain(
```
Comment (samuel-uniris) [inline lib/archethic/replication/transaction_context.ex:66]: Why add this new function ?

Comment (samuel-uniris) [reply lib/archethic/replication/transaction_context.ex:66]: You can leverage the existing one, no ?

File: lib/archethic/replication/transaction_context.ex [hunk 8]
```diff
@@ -42,19 +42,132 @@ defmodule ArchEthic.Replication.TransactionContext do
...
+         _prev_result \\ nil
+       ) do
+    case P2P.send_message(node, %GetTransactionChain{
+           address: address,
+           after: time_after,
+           page: page_state
+         }) do
+        # ------------------------------------------------------------------
+      #  case 0: erraneou cases
+      {:ok, %TransactionList{transactions: [], more?: _, page:  nil}} ->
```
Comment (samuel-uniris) [inline lib/archethic/replication/transaction_context.ex:99]: Why do you think it's an error when there are not transactions ?

File: lib/archethic/replication/transaction_context.ex [hunk 9]
```diff
@@ -42,19 +42,132 @@ defmodule ArchEthic.Replication.TransactionContext do
...
+        replicate_transaction_chain(rest, address, time_after, nil, [])
+
+      # ------------------------------------------------------------------
+      #  case 1: if  txn_chain have many transactions and more than one page of data and more page  is pending
+      #  then process transactions and request for next page data from same node against same time after
+      #  if any error look for same data and time_after from another nodes
+      {:ok, %TransactionList{transactions: transactions, more?: true, page: page}}  ->
+        case process_transactions(transactions, address, time_after, page) do
```
Comment (samuel-uniris) [inline lib/archethic/replication/transaction_context.ex:107]: I don't get the purpose of this call, if you don't handle the result

Comment (samuel-uniris) [reply lib/archethic/replication/transaction_context.ex:107]: Also you are not accumulating the transactions

**Review by samuel-uniris — CHANGES_REQUESTED** [review]

File: lib/archethic/replication/transaction_context.ex [hunk 10]
```diff
@@ -33,36 +33,38 @@ defmodule ArchEthic.Replication.TransactionContext do
...
 
       {:ok, %TransactionList{transactions: transactions, more?: false, page: _paging_state}} ->
-        process_transactions(transactions, address)
-        do_fetch_transaction_chain(rest, {address, time_after, nil})
-
-      # return txn or not?
+        do_fetch_transaction_chain(
+          rest,
+          {address, time_after, nil},
+          [List.flatten(transactions) | List.flatten(prev_result)] |> Enum.uniq()
```
Comment (samuel-uniris) [inline lib/archethic/replication/transaction_context.ex:63]: You do flatten two times, you can reduce it to one single for the concatenated list.

Comment (samuel-uniris) [reply lib/archethic/replication/transaction_context.ex:63]: I think you should return the results straight away. Otherwise, you will loop too many times for all the nodes.

**Review by samuel-uniris — CHANGES_REQUESTED** [review]

**Review by samuel-uniris — CHANGES_REQUESTED** [review]

File: lib/archethic/replication/transaction_context.ex [hunk 11]
```diff
@@ -33,27 +33,47 @@ defmodule ArchEthic.Replication.TransactionContext do
...
+    {address, time_after, page_state} = fetch_options
 
-      {:ok, %TransactionList{transactions: transactions}} ->
-        transactions
+    message = %GetTransactionChain{address: address, after: time_after, page: page_state}
+    # query all the nodes and keep uniqure txn only ends when no more nodes to query
+
+    case P2P.send_message(node, message) do
+      {:ok, %TransactionList{transactions: [], page: _}} ->
+        do_fetch_transaction_chain(rest, {address, time_after, page_state}, prev_result)
```
[the remark this answers is not in the document]
Comment (apoorv-2204) [reply lib/archethic/replication/transaction_context.ex:50]: Hi , maybe we can fetch from another node?? if other node have it?

File: lib/archethic/p2p/message/transaction_list.ex [hunk 12]
```diff
@@ -2,11 +2,12 @@ defmodule ArchEthic.P2P.Message.TransactionList do
...
   """
-  defstruct transactions: []
+  defstruct [:page, transactions: []]
 
   alias ArchEthic.TransactionChain.Transaction
 
   @type t :: %__MODULE__{
-          transactions: list(Transaction.t())
+          transactions: list(Transaction.t()),
+          page: nil | binary()
```
[the remark this answers is not in the document]
Comment (apoorv-2204) [reply lib/archethic/p2p/message/transaction_list.ex:11]: Hi, it was omitted due to, more? = page != nil

[the remark this answers is not in the document]
Comment (samuel-uniris) [reply lib/archethic/p2p/message/transaction_list.ex:11]: Page and more have not the same meaning and behavior.
Page serves to paginate data while More indicates if a node needs to request for more data.

File: lib/archethic.ex [hunk 13]
```diff
@@ -249,7 +249,7 @@ defmodule ArchEthic do
 
   defp get_transaction_chain([node | rest], address) do
     case P2P.send_message(node, %GetTransactionChain{address: address}) do
-      {:ok, %TransactionList{transactions: transactions}} ->
+      {:ok, %TransactionList{transactions: transactions, more?: false, page: nil}} ->
```
Comment (samuel-uniris) [inline lib/archethic.ex:252]: This might crash if you get `%TransactionList{more?: true, page: _}` due to unmatched pattern matching

File: lib/archethic/db/cassandra_impl.ex [hunk 14]
```diff
@@ -78,48 +78,100 @@ defmodule ArchEthic.DB.CassandraImpl do
...
   """
-  @spec get_transaction_chain(binary(), list()) :: Enumerable.t()
-  def get_transaction_chain(address, fields \\ []) when is_binary(address) and is_list(fields) do
+  @spec get_transaction_chain(binary(), list()) :: {list(Transaction.t()), boolean(), binary()}
+  def get_transaction_chain(address, all_fields \\ [])
+      when is_binary(address) and is_list(all_fields) do
     start = System.monotonic_time()
+    # order matter first query options should be passed first
+    # then field options should be passed
+    {options, fields} = Enum.split_while(all_fields, &get_options_and_fields?(&1))
```
Comment (samuel-uniris) [inline lib/archethic/db/cassandra_impl.ex:87]: I don't get this part, what is the link between options and fields.
i think it should be distinct.
Maybe pass an option parameter just for the query options

File: lib/archethic/oracle_chain/scheduler.ex [hunk 15]
```diff
@@ -414,7 +414,7 @@ defmodule ArchEthic.OracleChain.Scheduler do
   end
 
   defp send_summary_transaction(summary_date, index) do
-    oracle_chain =
+    {oracle_chain, _more?, _paging_state} =
```
Comment (samuel-uniris) [inline lib/archethic/oracle_chain/scheduler.ex:417]: I think it needs to loop recursively here also

File: lib/archethic/p2p/message.ex [hunk 16]
```diff
@@ -436,14 +448,24 @@ defmodule ArchEthic.P2P.Message do
...
 
-  def encode(%TransactionList{transactions: transactions}) do
+  def encode(%TransactionList{transactions: transactions, more?: false, page: nil}) do
+    transaction_bin =
+      transactions
+      |> Stream.map(&Transaction.serialize/1)
+      |> Enum.to_list()
+      |> :erlang.list_to_bitstring()
+
+    <<251::8, Enum.count(transactions)::32, transaction_bin::bitstring, "FALSE", "PAGE_NIL">>
```
Comment (samuel-uniris) [inline lib/archethic/p2p/message.ex:458]: I think you can encode in a more efficient way instead of "FALSE", you can use a bitstring: `0::1` , and opposite for "TRUE"

Comment (samuel-uniris) [inline lib/archethic/p2p/message.ex:458]: Also for "PAGE_NIL", you can maybe use: ""

File: lib/archethic_web/live/chains/oracle_live.ex [hunk 17]
```diff
@@ -170,10 +170,13 @@ defmodule ArchEthicWeb.OracleChainLive do
...
   defp list_transactions_by_date(date = %DateTime{}) do
-    date
-    |> Crypto.derive_oracle_address(0)
-    |> TransactionChain.get_last_address()
-    |> TransactionChain.get([:address, :type, validation_stamp: [:timestamp]])
+    {oracle_chain, _more?, _paging_state} =
+      date
+      |> Crypto.derive_oracle_address(0)
+      |> TransactionChain.get_last_address()
+      |> TransactionChain.get([:address, :type, validation_stamp: [:timestamp]])
```
Comment (samuel-uniris) [inline lib/archethic_web/live/chains/oracle_live.ex:177]: I think this should also be recursive here

File: lib/archethic/replication.ex [hunk 18]
```diff
@@ -226,7 +226,7 @@ defmodule ArchEthic.Replication do
       transaction_address: Base.encode16(tx.address)
     )
 
-    previous_chain = TransactionChain.get(previous_address)
+    {previous_chain, _more?, _paging_state} = TransactionChain.get(previous_address)
```
Comment (samuel-uniris) [inline lib/archethic/replication.ex:229]: I think this is a miss

**Review by samuel-uniris — CHANGES_REQUESTED** [review]

**Comment by samuel-uniris** [comment]
I got some errors after few transaction, regarding the chain integrity and download:

image

File: lib/archethic/db/cassandra_impl.ex [hunk 19]
```diff
@@ -128,7 +128,7 @@ defmodule ArchEthic.DB.CassandraImpl do
          address,
          after_time
        ) do
-    {" SELECT transaction_address FROM archethic.transaction_chains WHERE chain_address = ? AND transaction_timestamp >=  ? ",
+    {" SELECT transaction_address FROM archethic.transaction_chains WHERE chain_address = ? AND transaction_timestamp < ? ",
```
Comment (apoorv-2204) [inline lib/archethic/db/cassandra_impl.ex:132]: Hi, please <= ,
`{" SELECT transaction_address FROM archethic.transaction_chains WHERE chain_address = ? AND transaction_timestamp <= ? ",` It fixes docker testnet issues.

File: lib/archethic/p2p/message/get_transaction_chain.ex [hunk 20]
```diff
@@ -3,12 +3,13 @@ defmodule ArchEthic.P2P.Message.GetTransactionChain do
...
   @enforce_keys [:address]
-  defstruct [:address, :after]
+  defstruct [:address, :after, :paging_state]
 
   alias ArchEthic.Crypto
 
   @type t :: %__MODULE__{
           address: Crypto.versioned_hash(),
-          after: nil | DateTime.t()
+          after: nil | DateTime.t(),
```
Comment (apoorv-2204) [inline lib/archethic/p2p/message/get_transaction_chain.ex:12]: hi, please can we have before: ? its misleading.

**Comment by apoorv-2204** [comment]
`69, 108, 101, 210, 193, 160, 91, 147, 31, 50, 240, 251, 91, 166, 183, 81, 94, 4, 99, 208, 236, 58, 5>>}]
node2_1      | 2022-03-22 07:18:51.290 transaction_address=000068A9CC4F77F39C3CAAB4B38302B1ED61117BFD6F937FF9892599F5C78E893B4D transaction_type=transfer [error] Consensus not reached` hi we have a another issue of consensus not reaching. during docker testnet

File: lib/archethic/db/cassandra_impl.ex (hunk 19 above)
Comment (samuel-uniris) [reply lib/archethic/db/cassandra_impl.ex:132]: the previous state was: 
https://github.com/archethic-foundation/archethic-node/blob/c7b006a94e3621347161bfaee652fd7b9d8b4d6c/lib/archethic/p2p/message.ex#L1003

So I don't want to introduce something not testing properly

File: lib/archethic/p2p/message/get_transaction_chain.ex (hunk 20 above)
Comment (samuel-uniris) [reply lib/archethic/p2p/message/get_transaction_chain.ex:12]: Why do you want `?`

Comment (samuel-uniris) [reply lib/archethic/p2p/message/get_transaction_chain.ex:12]: Sorry I don't get, can you explain more ?
